mqtt/mqtt-broker/src/broker.rs

```diff
@@ -100,6 +104,64 @@ where
         Ok(self.snapshot())
     }
 
+    fn prepare_activities(client_id: &ClientId, session: &Session) -> Vec<Activity> {
+        let mut activities = Vec::new();
+        for sub in session
+            .subscriptions()
+            .into_iter()
+            .flat_map(HashMap::values)
+        {
+            if let Some(client_info) = session.client_info_omit_offline_sessions() {
+                let sub_topic_filter = sub.filter().to_string();
+                let operation = Operation::new_subscribe(proto::SubscribeTo {
+                    topic_filter: sub_topic_filter.clone(),
+                    qos: *sub.max_qos(),
+                });
+                activities.push(Activity::new(
+                    client_id.clone(),
+                    client_info.clone(),
+                    operation,
+                ));
+            }
+        }
+        activities
+    }
+
+    fn reevaluate_subscriptions(&mut self) -> Result<(), Error> {
+        let disconnecting: Vec<ClientId> = self
+            .sessions
+            .iter()
+            .flat_map(|(client_id, session)| Self::prepare_activities(client_id, session))
+            .filter_map(
+                |activity: Activity| match self.authorizer.authorize(activity.clone()) {
```

Comment:
This .clone() is the reason you added `#derive(Clone)` in a bunch of places, right?

Reply:
That's correct